lib/xenops_server.ml

```diff
@@ -1504,42 +1521,65 @@ and perform ?subtask ?result (op: operation) (t: Xenops_task.t) : unit =
-						Cohttp.Cookie.Cookie_hdr.serialize cookies :: [
-							"Connection", "keep-alive";
-							"User-agent", "xenopsd";
-						]) in
-					let request = Cohttp.Request.make ~meth:`PUT ~version:`HTTP_1_1 ~headers memory_url in
-
-					Request.write (fun _ -> ()) request mfd;
-
-					begin match Handshake.recv mfd with
-						| Handshake.Success -> ()
-						| Handshake.Error msg ->
-							error "cannot transmit vm to host: %s" msg;
-							raise (Internal_error msg)
-					end;
+				(fun mem_mfd ->
+					let do_request mfd memory_limit url vwhat = (
+						Sockopt.set_sock_keepalives mfd;
+						let open Xenops_migrate in
+						let module Request = Cohttp.Request.Make(Cohttp_posix_io.Unbuffered_IO) in
+						let cookies = [
+							"instance_id", instance_id;
+							"dbg", t.Xenops_task.dbg;
+							"memory_limit", Int64.to_string memory_limit;
+						] in
+						let headers = Cohttp.Header.of_list (
+							Cohttp.Cookie.Cookie_hdr.serialize cookies :: [
+								"Connection", "keep-alive";
+								"User-agent", "xenopsd";
+							]) in
+						let request = Cohttp.Request.make ~meth:`PUT ~version:`HTTP_1_1 ~headers url in
+
+						Request.write (fun _ -> ()) request mfd;
+
+						begin match Handshake.recv mfd with
+							| Handshake.Success -> ()
+							| Handshake.Error msg ->
+								error "cannot transmit v%s to host: %s" vwhat msg;
+								raise (Internal_error msg)
+						end;
+					) in
+
+					do_request mem_mfd state.Vm.memory_limit memory_url "m";
 					debug "Synchronisation point 1";
 
-					perform_atomics [
-						VM_save(id, [ Live ], FD mfd)
-					] t;
-					debug "Synchronisation point 2";
-
-					Handshake.send ~verbose:true mfd Handshake.Success;
-					debug "Synchronisation point 3";
-
-					Handshake.recv_success mfd;
-					debug "Synchronisation point 4";
+					let save_then_handshake atom mfd p1 p2 p3 = (
+						perform_atomics [atom] t;
+						debug "Synchronisation point %s" p1;
+
+						let module Handshake = Xenops_migrate.Handshake in
+						Handshake.send ~verbose:true mfd Handshake.Success;
+						debug "Synchronisation point %s" p2;
+
+						Handshake.recv_success mfd;
+						debug "Synchronisation point %s" p3;
+					) in
+					let save_vm_then_handshake () =
+						save_then_handshake (VM_save (id, [ Live ], FD mem_mfd)) mem_mfd "2" "3" "4" in
+
+					match VGPU_DB.ids id with
+					| [] ->
+						save_vm_then_handshake ()
+					| [vgpu_id] ->
+						let vgpu_url = make_url "/migrate-vgpu/" in
+						let vgpu_memory_limit = 1024L in (* TODO find this out from vgpu_id (or vgpu if use use VGPU_DB.vgpus) *)
+(*						let vgpu_state = B.VGPU.get_state id in (* pass the id of the VM *)
+						let vgpu_memory_limit = vgpu_state.Vgpu.fb_len in *)
+						Open_uri.with_open_uri vgpu_url (fun vgpu_mfd ->
+							do_request vgpu_mfd vgpu_memory_limit vgpu_url "gpu";
+							debug "Synchronisation point 1.2";
+							save_then_handshake (VGPU_save (vgpu_id, FD vgpu_mfd)) vgpu_mfd "1.4" "1.6" "1.8";
```

Comment:
Too much handshaking?
In the no-vgpu case it's the same, and if we have a vgpu then we handshake for the memory (as always) and do the same for the vgpu. What should we do instead?